### ISSUE TITLE
Running ansible-lint against a single file reduces different results from running it against the entire playbook

### ISSUE THREAD

**Issue by mhogerheijde** [ansible/ansible-lint#280]
# Issue Type
- Bug report
or
- Feature request

Depending on whether or not this is intended behaviour

# Ansible and Ansible Lint details

```
$ ansible --version
ansible 2.3.2.0
  config file = /home/mhogerheijde/workspace/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
$ ansible-lint --version
ansible-lint 3.4.15

```

- ansible installation method: pip
- ansible-lint installation method: pip

# Desired Behaviour

When running ansible-lint against a single file, I expect the same report as when I run ansible-lint against the entire playbook.

# Actual Behaviour (Bug report only)

When running ansible-lint against the entire playbook (site.yml), errors are reported (as expected). When running ansible-lint against the file containing issues, no error are reported.

```
$ ansible-lint -R /home/mhogerheijde/workspace/ansible/site.yml
[ANSIBLE0011] All tasks should be named
/home/mhogerheijde/workspace/ansible/roles/common/tasks/main.yml:26
Task/Handler: stat __file__=/home/mhogerheijde/workspace/ansible/roles/common/tasks/main.yml __line__=26 path=/etc/apt/apt.conf.d/20auto-upgrades

[ANSIBLE0011] All tasks should be named
/home/mhogerheijde/workspace/ansible/roles/common/tasks/main.yml:39
Task/Handler: locale_gen state=present name={{ item }} __file__=/home/mhogerheijde/workspace/ansible/roles/common/tasks/main.yml __line__=39

$
```

Versus running it against the specific file that contains the issues:

```
$ ansible-lint -R /home/mhogerheijde/workspace/ansible/roles/common/tasks/main.yml
$
```

----

It might look like this is related to either:

 - https://github.com/willthames/ansible-lint/issues/210
 - https://github.com/willthames/ansible-lint/issues/247
 - https://github.com/willthames/ansible-lint/issues/193

but not entirely.

I want to be able to (quickly) lint a single file and be able to no validate an entire playbook. So validating the role is not what I want to do. (The Atom linter will lint only one file at the time. I know you don't have a dependency on the Atom editor, but this shows a use-case.)

**Comment by jamescooke** [comment]
I'm also having a similar problem - I want to be able to lint a list of files, but when I call `ansible-lint` on each of those files in turn, no errors are returned.

I've found that when I make the `ansible-lint` call with the `-v` flag is shows me:

* When calling on a directory that `ansible-lint` is processing the file "of type tasks" and errors appear.

    ```sh
    $ ansible-lint -v roles/deploy-api/
    Examining /[full path]/roles/deploy-api/tasks/main.yml of type tasks
    [ANSIBLE0004] Git checkouts must contain explicit version
    /[full path]/roles/deploy-api/tasks/main.yml:2
    Task/Handler: checkout the server repo
    ...
    ```
* When calling on a file that `ansible-lint` is processing the file "of type playbook" and **no** errors appear!

    ```sh
    $ ansible-lint -v roles/deploy-api/tasks/main.yml
    Examining roles/deploy-api/tasks/main.yml of type playbook
    $
    ```

**Comment by jamescooke** [comment]
Also note as per #210 , missing the trailing slash on `$ ansible-lint -v roles/deploy-api/` will mean that no files are checked.

**Comment by p4ulie** [comment]
Similar experience here.
Only file tmp/tasks/main.yml is getting evaluated as "type task" and produces correct output when running "ansible-lint -v tmp": 

```
bash-3.2$ tree tmp
tmp
|-- roles
|   `-- main.yml
`-- tasks
    `-- main.yml
2 directories, 2 files
```

```
bash-3.2$ ansible-lint -v tmp/
```

```
bash-3.2$ ansible-lint -v tmp/tasks/main.yml
```

And finally:
```
bash-3.2$ ansible-lint -v tmp
Examining /Users/paulie/tmp/tasks/main.yml of type tasks
[ANSIBLE0012] Commands should not change things if nothing needs doing
/Users/paulie/tmp/tasks/main.yml:1
Task/Handler: this would typically fire CommandsInsteadOfArgumentRule

[ANSIBLE0012] Commands should not change things if nothing needs doing
/Users/paulie/tmp/tasks/main.yml:4
Task/Handler: this would typically fire CommandsInsteadOfModuleRule

[ANSIBLE0004] Git checkouts must contain explicit version
/Users/paulie/tmp/tasks/main.yml:9
Task/Handler: this would typically fire GitHasVersionRule
```

**Comment by willthames** [comment]
To be honest, ansible-lint wasn't really designed for processing individual files other than playbooks - it gets a lot of its context from how things are included from a playbook - if it just gets asked to process a yaml file it will indeed assume it's a playbook, as @jamescooke notes.

[ansible-review](https://github.com/willthames/ansible-review) does a better job of determining file types and is designed to process individual files, with better control over which checks are run.

**Comment by jamescooke** [comment]
@willthames Thanks for the clarification. I wasn't aware of ansible-review.

For now I'll continue to carefully manage the list of files and folders passed to ansible-lint in CI configs. But maybe there's some value in a warning being displayed? Although I'm not sure what the rules for that would be.

**Comment by mhogerheijde** [comment]
@willthames Thanks for the clarification! 

So basically the configuration of the Atom linter is incorrect? (It should either call ansible-lint with the entire playbook, (and/)or call ansible-review for single files.

**Comment by ssbarnea** [comment]
Identical results cannot be guaranteed but with newer versions I doubt you will face the same problems that you originally had.

**Comment by paulkitt** [comment]
No this is still an issue running ansible-lint 4.3.4 and ansible 2.9.13 on Arch Linux!
The output when executing ansible-lint on a whole playbook or folder differs still significantly from running it against a single file:

```
$ ansible-lint -vp roles/deploy_experimental_app
roles/deploy_experimental_app/tasks/main.yml:10: [E201] Trailing whitespace
roles/deploy_experimental_app/tasks/main.yml:13: [E401] Git checkouts must contain explicit version
roles/deploy_experimental_app/tasks/main.yml:20: [E201] Trailing whitespace
roles/deploy_experimental_app/tasks/nginx_config.yml:3: [E208] File permissions not mentioned
roles/deploy_experimental_app/tasks/nginx_config.yml:4: [E201] Trailing whitespace
```

```
$ ansible-lint -vp roles/deploy_experimental_app/tasks/nginx_config.yml                              
roles/deploy_experimental_app/tasks/nginx_config.yml:4: [E201] Trailing whitespace
```
Please reopen this issue!

**Comment by ssbarnea** [comment]
As I mentioned earlier, that is expected due to the nature of the processing, it is context dependent and when passing specific files the "coverage" is reduced.

I am not going to reopen the ticket but you are welcomed to propose a PR that addresses the issue. Closed tickets does not mean that PRs are not welcomed.

**Comment by paulkitt** [comment]
Ok to bad! Thanks for the fast clarification. Running ansible-lint inside an editor will then miss certain problems and we need to run linting the whole playbook/project in our pipeline.

**Comment by ssbarnea** [comment]
~9/10 rules should work at individual file level, but the others would not. Also the number of contextual checks is likely to be growing.
